### PR TITLE
Feat: helper script to read KeyVault secrets

### PR DESCRIPTION
**Pull request by thekaveman**
Part of #1847 

## What this PR does

* Defines a new setting `RUNTIME_ENVIRONMENT()` to dynamically calculate one of `local`, `dev`, `test`, or `prod` based on the values in `ALLOWED_HOSTS`
* Uses [`azure-keyvault-secrets`](https://pypi.org/project/azure-keyvault-secrets/) and [`azure-identity`](https://pypi.org/project/azure-identity/) based on [Getting Started with Azure Key Vault Secrets client library for Python](https://learn.microsoft.com/en-us/python/api/overview/azure/keyvault-secrets-readme?view=azure-python&source=recommendations#getting-started)
* Installs the Azure CLI into the devcontainer, to make debugging and testing of this functionality possible within the same environment we use for daily development.
* For `local` runtime environment, falls back to reading from environment variables

## Todo

- [x] Figure out how (ideally via Terraform) to create a [Managed identity for App Services](https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Cpython) so this code can run in Azure (non-local) environments. See also https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/identity/azure-identity#managed-identity-support

### Done

The existing Terraform [`app_service` definition](https://github.com/cal-itp/benefits/blob/dev/terraform/app_service.tf#L36) includes a [system-defined managed identity](https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Cpython).

Further, this managed identity should be able to GET secrets via an existing [policy definition](https://github.com/cal-itp/benefits/blob/dev/terraform/key_vault.tf#L77).

[According to the Python package docs](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/identity/azure-identity#defaultazurecredential), `DefaultCredential` will check for access via this system-defined managed identity when running in Azure. This PR adds a `DEBUG`-only route `/testsecret` to help verify access from within Azure.

## How to verify the POC

1. Update your local `.env` file to add a new variable

        testsecret=Hello from the local environment!

1. Rebuild and Reopen the devcontainer

1. Call the helper script, setting the `DJANGO_SETTINGS_MODULE` env var and passing the name of the secret to read

        DJANGO_SETTINGS_MODULE=benefits.settings \
        python benefits/secrets.py testsecret

1. Verify output:
 
        [local] testsecret: Hello from the local environment!

1. From the devcontainer terminal, login to the Azure CLI with the `--use-device-code` flag, following instructions printed in the terminal. You may need to login to Azure on the website first, if you receive an error related to 2FA, to trigger that flow. Then run:
        
        az login --use-device-code

1. Call the helper script again configuring `DJANGO_ALLOWED_HOSTS` for the `dev` environment:

        DJANGO_ALLOWED_HOSTS=dev-benefits.calitp.org \
        DJANGO_SETTINGS_MODULE=benefits.settings \
        python benefits/secrets.py testsecret
        
1. Verify output:
 
        [dev] testsecret: Hello from the dev KeyVault!

1. Launch the local app with `F5` and visit the `/testsecret` endpoint, verify output in the browser:

        Hello from the local environment!

1. Override `DJANGO_ALLOWED_HOSTS` in `.vscode/launch.json` to point to `dev`. Be sure to include `localhost` so you can launch the app!

        "env": {
          ...
          "DJANGO_ALLOWED_HOSTS": "dev-benefits.calitp.org,localhost",
        }

1.  Launch the local app with `F5` and visit the `/testsecret` endpoint, verify output in the browser:

        Hello from the dev KeyVault!

1. Try another call against e.g. the `test` environment:

        DJANGO_ALLOWED_HOSTS=test-benefits.calitp.org \
        DJANGO_SETTINGS_MODULE=benefits.settings \
        python benefits/secrets.py <another secret name>

1. And verify the expected output:

        [test] <secret name>: <secret value from test>

## Post-merge verification steps

After this PR is merged and deployed to `dev`, visit the `/testsecret` route, which should print the value of the test secret in the browser (shown in `localhost` below):

image

### CODE REVIEW

File: benefits/secrets.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        exit(1)
+
+    vault_url = args[0]
+    secret_name = args[1]
+
+    credential = DefaultAzureCredential()
+    client = SecretClient(vault_url=vault_url, credential=credential)
+    secret = client.get_secret(secret_name)
+
+    print(f"Reading {secret_name} from {vault_url}")
```
[the remark this answers is not in the document]
Comment (thekaveman) [reply benefits/secrets.py:20]: This is a POC, not production code

File: benefits/settings.py [hunk 2]
```diff
@@ -24,6 +26,20 @@
...
 ALLOWED_HOSTS = _filter_empty(os.environ.get("DJANGO_ALLOWED_HOSTS", "localhost,127.0.0.1").split(","))
 
+
+def RUNTIME_ENVIRONMENT():
+    """Helper calculates the current runtime environment from ALLOWED_HOSTS."""
+
+    # usage of django.conf.settings.ALLOWED_HOSTS here (rather than the module variable directly)
+    # is to ensure dynamic calculation, e.g. for unit tests and elsewhere this setting is needed
+    env = "dev"
+    if "test-benefits.calitp.org" in settings.ALLOWED_HOSTS:
```
[the remark this answers is not in the document]
Comment (thekaveman) [reply benefits/settings.py:36]: > Sanitizing untrusted URLs is a common technique for preventing attacks such as request forgeries and malicious redirections. Usually, this is done by checking that the host of a URL is in a set of allowed hosts.

This is literally what this code is doing... ?

**Comment by angela-tran** [comment]
### Verifying the POC

I was able to go through steps 1-4 and 6 without any issue. ✅ 

For step 5, I was able to go to `/testsecret` locally and see the secret in my browser. 


Just curious though: do you see these errors in the terminal in VS Code?
```
[02/Feb/2024 21:05:33] DEBUG benefits.core.session:251 Update session debug
[02/Feb/2024 21:05:33] INFO azure.identity._credentials.environment:109 No environment configuration found.
[02/Feb/2024 21:05:33] INFO azure.identity._credentials.managed_identity:96 ManagedIdentityCredential will use IMDS
[02/Feb/2024 21:05:33] DEBUG azure.identity._internal.decorators:53 EnvironmentCredential.get_token failed: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.
Visit https://aka.ms/azsdk/python/identity/environmentcredential/troubleshoot to troubleshoot this issue.
Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/decorators.py", line 33, in wrapper
    token = fn(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_credentials/environment.py", line 150, in get_token
    raise CredentialUnavailableError(message=message)
azure.identity._exceptions.CredentialUnavailableError: EnvironmentCredential authentication unavailable. Environment variables are not fully configured.
Visit https://aka.ms/azsdk/python/identity/environmentcredential/troubleshoot to troubleshoot this issue.
[02/Feb/2024 21:05:33] INFO azure.core.pipeline.policies.http_logging_policy:514 Request URL: 'http://169.254.169.254/metadata/identity/oauth2/token?api-version=REDACTED&resource=REDACTED'
Request method: 'GET'
Request headers:
    'User-Agent': 'azsdk-python-identity/1.15.0 Python/3.11.6 (Linux-6.5.11-linuxkit-x86_64-with-glibc2.36)'
No body was attached to the request
[02/Feb/2024 21:05:33] DEBUG urllib3.connectionpool:244 Starting new HTTP connection (1): 169.254.169.254:80
[02/Feb/2024 21:05:33] DEBUG urllib3.connectionpool:546 http://169.254.169.254:80 "GET /metadata/identity/oauth2/token?api-version=2018-02-01&resource=https://vault.azure.net HTTP/1.1" 403 None
[02/Feb/2024 21:05:33] INFO azure.core.pipeline.policies.http_logging_policy:550 Response status: 403
Response headers:
    'Connection': 'close'
[02/Feb/2024 21:05:33] DEBUG charset_normalizer:439 Encoding detection: ascii is most likely the one.
[02/Feb/2024 21:05:33] DEBUG azure.identity._internal.get_token_mixin:101 ImdsCredential.get_token failed: ManagedIdentityCredential authentication unavailable, no response from the IMDS endpoint.
Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/core/pipeline/policies/_universal.py", line 614, in deserialize_from_text
    return json.loads(data_as_str)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/managed_identity_client.py", line 41, in _process_response
    content = ContentDecodePolicy.deserialize_from_text(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/core/pipeline/policies/_universal.py", line 616, in deserialize_from_text
    raise DecodeError(
azure.core.exceptions.DecodeError: JSON is invalid: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_credentials/imds.py", line 87, in _request_token
    self._client.request_token(*scopes, connection_timeout=1, retry_total=0)
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/managed_identity_client.py", line 113, in request_token
    token = self._process_response(response, request_time)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/managed_identity_client.py", line 45, in _process_response
    if response.http_response.content_type.startswith("application/json"):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'startswith'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/get_token_mixin.py", line 86, in get_token
    token = self._request_token(*scopes, claims=claims, tenant_id=tenant_id, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_credentials/imds.py", line 97, in _request_token
    raise CredentialUnavailableError(error_message) from ex
azure.identity._exceptions.CredentialUnavailableError: ManagedIdentityCredential authentication unavailable, no response from the IMDS endpoint.
[02/Feb/2024 21:05:33] DEBUG azure.identity._internal.decorators:53 ManagedIdentityCredential.get_token failed: ManagedIdentityCredential authentication unavailable, no response from the IMDS endpoint.
Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/core/pipeline/policies/_universal.py", line 614, in deserialize_from_text
    return json.loads(data_as_str)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/__init__.py", line 346, in loads
    return _default_decoder.decode(s)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/managed_identity_client.py", line 41, in _process_response
    content = ContentDecodePolicy.deserialize_from_text(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/core/pipeline/policies/_universal.py", line 616, in deserialize_from_text
    raise DecodeError(
azure.core.exceptions.DecodeError: JSON is invalid: Expecting value: line 1 column 1 (char 0)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_credentials/imds.py", line 87, in _request_token
    self._client.request_token(*scopes, connection_timeout=1, retry_total=0)
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/managed_identity_client.py", line 113, in request_token
    token = self._process_response(response, request_time)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/managed_identity_client.py", line 45, in _process_response
    if response.http_response.content_type.startswith("application/json"):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'startswith'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/decorators.py", line 33, in wrapper
    token = fn(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_credentials/managed_identity.py", line 137, in get_token
    return self._credential.get_token(*scopes, claims=claims, tenant_id=tenant_id, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/get_token_mixin.py", line 86, in get_token
    token = self._request_token(*scopes, claims=claims, tenant_id=tenant_id, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_credentials/imds.py", line 97, in _request_token
    raise CredentialUnavailableError(error_message) from ex
azure.identity._exceptions.CredentialUnavailableError: ManagedIdentityCredential authentication unavailable, no response from the IMDS endpoint.
[02/Feb/2024 21:05:33] DEBUG azure.identity._persistent_cache:105 msal-extensions is unable to encrypt a persistent cache: "Unable to import module 'gi'
Runtime dependency of PyGObject is missing.
Depends on your Linux distro, you could install it system-wide by something like:
    sudo apt install python3-gi python3-gi-cairo gir1.2-secret-1
If necessary, please refer to PyGObject's doc:
https://pygobject.readthedocs.io/en/latest/getting_started.html
"
Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/msal_extensions/libsecret.py", line 18, in <module>
    import gi  # https://github.com/AzureAD/microsoft-authentication-extensions-for-python/wiki/Encryption-on-Linux  # pylint: disable=line-too-long
    ^^^^^^^^^
ModuleNotFoundError: No module named 'gi'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_persistent_cache.py", line 101, in _get_persistence
    return msal_extensions.LibsecretPersistence(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/msal_extensions/persistence.py", line 314, in __init__
    from .libsecret import (  # This uncertain import is deferred till runtime
  File "/home/calitp/.local/lib/python3.11/site-packages/msal_extensions/libsecret.py", line 20, in <module>
    raise ImportError("""Unable to import module 'gi'
ImportError: Unable to import module 'gi'
Runtime dependency of PyGObject is missing.
Depends on your Linux distro, you could install it system-wide by something like:
    sudo apt install python3-gi python3-gi-cairo gir1.2-secret-1
If necessary, please refer to PyGObject's doc:
https://pygobject.readthedocs.io/en/latest/getting_started.html

[02/Feb/2024 21:05:33] DEBUG azure.identity._internal.decorators:53 SharedTokenCacheCredential.get_token failed: SharedTokenCacheCredential authentication unavailable. No accounts were found in the cache.
Traceback (most recent call last):
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/decorators.py", line 33, in wrapper
    token = fn(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_credentials/shared_cache.py", line 80, in get_token
    return self._credential.get_token(*scopes, claims=claims, tenant_id=tenant_id, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_credentials/shared_cache.py", line 124, in get_token
    account = self._get_account(self._username, self._tenant_id, is_cae=is_cae)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/decorators.py", line 79, in wrapper
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/calitp/.local/lib/python3.11/site-packages/azure/identity/_internal/shared_token_cache.py", line 205, in _get_account
    raise CredentialUnavailableError(message=NO_ACCOUNTS)
azure.identity._exceptions.CredentialUnavailableError: SharedTokenCacheCredential authentication unavailable. No accounts were found in the cache.
[02/Feb/2024 21:05:33] DEBUG azure.identity._internal.decorators:34 AzureCliCredential.get_token succeeded
[02/Feb/2024 21:05:33] DEBUG azure.identity._internal.decorators:48 [Authenticated account] Client ID: <ID>. Tenant ID: <ID>. User Principal Name: unavailableUpn. Object ID (user): <ID>
[02/Feb/2024 21:05:33] INFO azure.identity._credentials.chained:95 DefaultAzureCredential acquired a token from AzureCliCredential
[02/Feb/2024 21:05:33] INFO azure.core.pipeline.policies.http_logging_policy:514 Request URL: 'https://<KEY VAULT URL>/secrets/testsecret/?api-version=REDACTED'
Request method: 'GET'
Request headers:
    'Accept': 'application/json'
    'x-ms-client-request-id': 'ce000502-c20e-11ee-b4dc-0242ac150004'
    'User-Agent': 'azsdk-python-keyvault-secrets/4.7.0 Python/3.11.6 (Linux-6.5.11-linuxkit-x86_64-with-glibc2.36)'
    'Authorization': 'REDACTED'
No body was attached to the request
```

**Comment by thekaveman** [comment]
@angela-tran 

> Just curious though: do you see these errors in the terminal in VS Code?

Yes! I hadn't noticed these before, but I do see them in trying to hit the local `/testsecret` endpoint. 

I think this reflects [`azure-identity`'s stated strategy](https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/identity/azure-identity#defaultazurecredential) for how `DefaultAzureCredential` attempts to resolve:

> `DefaultAzureCredential` attempts to authenticate via the following mechanisms, in this order, stopping when one succeeds...
> ...
> As of version 1.14.0, `DefaultAzureCredential` will attempt to authenticate with all developer credentials until one succeeds, regardless of any errors previous developer credentials experienced. For example, a developer credential may attempt to get a token and fail, so `DefaultAzureCredential` will continue to the next credential in the flow.

Since these are `DEBUG` level messages, and because it follows the published docs, I'm not too worried about it.

**Review by angela-tran — DISMISSED** [review]

**Comment by thekaveman** [comment]
@machikoyasuda please make sure you can go through the [`How to Verify` steps above](https://github.com/cal-itp/benefits/pull/1859#issue-2102856340) before we merge this.

**Comment by thekaveman** [comment]
Moving back to draft as I work on some additional fixes for running the app locally and unit/integration testing.

**Comment by thekaveman** [comment]
@angela-tran @machikoyasuda ready for re-review! Please see the updated steps in [How to verify the POC](https://github.com/cal-itp/benefits/pull/1859#issue-2102856340) above.

**Review by angela-tran — APPROVED** [review]
I was able to confirm support for local configuration

**Review by machikoyasuda — APPROVED** [review]
Able to go thru all the testing steps and read KV secret values for local, dev and test

**Pull request merged**